### DISCUSSION TITLE
Keep loop running

### DISCUSSION THREAD

**Discussion by mdkcore0** [libuv/libuv#3318] [Q&A]
Hi, I'm a newcomer to libuv, and it's great! Many thanks for all your work!

I'm migrating an application (it uses Qt and its event loop); the application is pretty simple now: should register some callbacks (the callbacks are against a so file only, that the device vendor gave to us :/), and listen/react to those events. It handles some specific usb devices and events.

I already have it working, but have a question about keeping the uv loop running (since the app quits after initializing the callbacks), and tried 3 solutions that seems to work (minimal code):

A:
```
bool running = true; // is set to false on a SIGINT callback
int main()
{
    library_initialization(callback1, callback2);
    ...
    int ret = uv_run(loop, UV_RUN_DEFAULT);
    while (running) {
        std::this_thread::sleep_for(std::chrono::milliseconds(1));
    }
    return ret;
}
```
B:
```
bool running = true; // is set to false on a SIGINT callback

void run(uv_idle_t *handle)
{
    std::this_thread::sleep_for(std::chrono::milliseconds(1));
}

int main()
{
    library_initialization(callback1, callback2);
    ...
    uv_idle_t idle;
    uv_idle_init(loop, &idle);
    uv_idle_start(&idle, run);
    return uv_run(loop, UV_RUN_DEFAULT);
}
```
C:
```
bool running = true; // is set to false on a SIGINT callback

int main()
{
    library_initialization(callback1, callback2);
    ...
    int ret = 0;
    while (running) {
        ret = uv_run(loop, UV_RUN_ONCE);
    }
    return ret;
}
```
All of them works as expected. Don't like so much **A** approach, **B** seems an wrong use of _uv_idle_t_, and **C** is what I'm using now.

Any consideration, or recommendation of alternatives, on how I'm using libuv?
Thanks!

EDIT: just found out that handling a _uv_signal_t_ attached to SIGINT,  I can skip controlling running state, doing the proper application cleanup and using UV_RUN_DEFAULT as well. This seems to be the clear approach until now.

**Comment by JiuJiu119** [comment]
hi, 
how do you  get it work as Qt has it's own event loop?

**Reply by mdkcore0** [reply]
Hi, I did not tried to run them together, I was just porting an existing demo application that uses Qt to another one without the use of Qt
